GH-15614 enable h2o frame to pandas frame conversion using multi-thread from datatable [nocheck]

wendycwong opened the pull request:
This PR implements issue in GH: https://github.com/h2oai/h2o-3/issues/15614

Implemented workaround suggested by @meganjkurka  with code snippet from @Bernard-H2O.  Added python test to show that speedup actually occurred.

mn-mikke commented on h2o-py/h2o/frame.py:
```diff
@@ -1929,13 +1929,15 @@ def structure(self):
         :param bool use_pandas: If True (default) then return the H2OFrame as a pandas DataFrame (requires that the
             ``pandas`` library was installed). If False, then return the contents of the H2OFrame as plain nested
             list, in a row-wise order.
+        :param bool use_databale: if True and if use_pandas is True, will use datatable to speedup the conversion from
```
typo: `use_databale`

wendycwong replied:
Changed

mn-mikke commented on h2o-py/tests/testdir_misc/pyunit_gh_15614_datatable_2_pandas.py:
```diff
@@ -0,0 +1,32 @@
+# if datable is installed, this test will show that using datatable to convert h2o frame to pandas frame is
+# much faster.  This workaround is suggested by Megan Kurka.  I implemented the code snippet from Bernard Ong.
+def test_datatable_pandas():
+    if can_use_pandas() and can_use_datatable():    # only perform tests if pandas and datatable are installed
```
Do we have `datatable` installed in our python testing images?

wendycwong replied:
Not sure.  Is this something I can ask the DevOps team to do?  I fixed the typo.

meganjkurka commented on h2o-py/h2o/frame.py:
```diff
@@ -1954,6 +1956,17 @@ def as_data_frame(self, use_pandas=True, header=True):
             import pandas
+            if use_datatable:
+                if not(can_use_datatable()):
+                    warnings("Please install datatable to take adavantage of the speedup.  Default to original "
```
type - `adavantage`

wendycwong replied:
Changed.

tomasfryda commented:
I have conflicting feelings about this PR. It has 2 reasons:

1) if we want to use different library than `pandas`, should we use `datatable`? 
There is at least one library that seems better to me both in terms of popularity ([datatable](https://pypistats.org/packages/datatable) vs [polars](https://pypistats.org/packages/polars)) and CSV import speed - see the following table: measurements are in seconds, measurements are done on my macbook using 2.6GB CSV (`smalldata/wa_cannabis/raw/Dashboard_Usable_Sales_w_Weight_Daily.csv`)

![Screen Shot 2023-07-03 at 10 17 00](https://github.com/h2oai/h2o-3/assets/61695433/45ff4a0e-ae6f-4175-a219-6995da017017)

2) should we use CSV for "IPC"? There are other formats that were designed with this use-case in mind (e.g. [feather](https://arrow.apache.org/docs/python/feather.html)). I know Apache Arrow only vaguely but IIRC the feather file format is basically a memory dump of the arrow's data structures so it's very fast and since both Wes McKinney (`pandas`' author) and Hadley Wickham (major R contributor (e.g. `ggplot2`, `tidyverse`,..)) [collaborated on the creation of the feather file format](https://posit.co/blog/feather/) the support is relatively good. You can even notice that in my little "benchmark" the best solution is to use `polars` with `pyArrow` which is a bit of cheating because in this scenario `pandas` uses [arrow arrays and not `numpy` arrays](https://pola-rs.github.io/polars/py-polars/html/reference/dataframe/api/polars.DataFrame.to_pandas.html) so there is no need to copy the data and if there is some operation that is not supported by arrow compute then it will be lazily converted to `numpy` arrays but it also demonstrates the fact that pandas can use arrow under the hood.

What's more the Apache Arrow provides a [java library](https://arrow.apache.org/docs/java/ipc.html) as well.

The most recent benchmark comparing various formats: https://rsangole.netlify.app/posts/2022-09-14_data-read-write-performance/data-read-write-perf

Note that even if this section might sound like I'd suggest to go ahead and implement arrow/feather support in H2O, that's not what I wanted to convey. All I was trying to say is that I think there are better ways of accomplishing faster H2OFrame to pandas' DataFrame conversion. 

----

With all that being said I also have a suggestion for this PR: I would like to have a test that would verify that the data types are same in both cases (`h2o->pandas` and `h2o->datatable->pandas`).

Since we use a CSV that doesn't explicitly mention the data type there could be issues like integers being converted to floats, datetime-related issues (will it still be a datetime or will it get converted to long, will it use the same timezone or will it end up without timezone etc), factor-related issues (will it be represented as factor (`pandas.Categorical`) will the levels have the same order), will it behave the same way when I ask for just one row from H2OFrame and the row will contain NA (fallback to the same type) etc.

I know this might sound like unnecessary work but once we'll have `datatable` installed it will be tested with `datatable` and I assume we don't want to duplicate all the tests to include both scenarios with and without `datatable` so we should have at least a basic confidence that it will work the same way even without `datatable`.

This could make testing visualizations especially painful - will the test notice if on the plot we have an index of the level of a factor instead of the level itself? (Unfortunately no, I have some automation locally that I use to test explain plots but since they end up different with different matplotlib versions it's not completely automatic (e.g., slight change layout engine changes a lot of pixels) so I have to visually inspect every single plot (but since we don't support Python2 now maybe it could be automated more easily))

wendycwong commented:
I have added polars as another way to convert h2o frame to pandas frame.  Also added tests to test out both ways.  I added comparison of both frames to make sure the comparison is fair.

tomasfryda commented on h2o-py/tests/testdir_misc/pyunit_gh_15614_datatable_2_pandas.py:
```diff
@@ -0,0 +1,43 @@
+        print("H2O frame to Pandas frame conversion time: {0}".format(oldTime))
+        # convert frame using datatable
+        startT = time.time()
+        new_pandas_frame = h2oFrame.as_data_frame(multi_thread=True, use_datatable=True)
```
This will likely fail due to `use_datatable` keyword argument - it's not implemented anymore. I'd remove this test.

tomasfryda commented on h2o-py/tests/testdir_misc/pyunit_gh_15614_polars_2_pandas.py:
```diff
@@ -0,0 +1,40 @@
+
+# if polars is installed, this test will show that using polars to convert h2o frame to pandas frame is
+# much faster.
+def test_datatable_pandas():
```
```suggestion
def test_polars_pandas():
```

tomasfryda commented on h2o-py/tests/testdir_misc/pyunit_gh_15614_polars_2_pandas.py:
```diff
@@ -0,0 +1,40 @@
+
+
+if __name__ == "__main__":
+    pyunit_utils.standalone_test(test_datatable_pandas)
```
```suggestion
    pyunit_utils.standalone_test(test_polars_pandas)
```

tomasfryda commented on h2o-py/tests/testdir_misc/pyunit_gh_15614_polars_2_pandas.py:
```diff
@@ -0,0 +1,40 @@
+if __name__ == "__main__":
+    pyunit_utils.standalone_test(test_datatable_pandas)
+else:
+    test_datatable_pandas()
```
```suggestion
    test_polars_pandas()
```

tomasfryda commented:
I would add 2 more datasets to the test so we know if that it doesn't change the behavior of the `as_data_frame` method:
- `smalldata/titanic/titanic_expanded.csv` - contains missing values, categorical and string variables
- `smalldata/timeSeries/CreditCard-ts_train.csv` - contains dates, integers, categorical variables and binary but numerical (`0`,`1`) variable

wendycwong commented:
I removed the datatable test and added two more dataset to the pyunit test.

tomasfryda reviewed (approved):
Looks good to me, thank you @wendycwong ! 

NOTE: some tests are failing so it would be good to rerun them.
